src/Apropos/HasPermutationGenerator/Abstraction.hs

```diff
@@ -35,12 +46,39 @@ abstract abstraction edge =
           (abstractionName abstraction)
           (propertyAbstraction abstraction)
           $ contract edge
-    , morphism = \m -> do
-        let n = m ^. modelAbstraction abstraction
-        nn <- morphism edge n
-        pure $ modelAbstraction abstraction .~ nn $ m
+    , morphism = productModelAbstraction abstraction $ morphism edge
+    }
+abstract abstraction@SumAbstraction {} edge =
+  Morphism
+    { name = abstractionName abstraction <> " of ( " <> name edge <> " )"
+    , match = Var (propLabel abstraction) :&&: ((propertyAbstraction abstraction #) <$> match edge)
+    , contract =
+        abstractContract
+          (abstractionName abstraction)
+          (propertyAbstraction abstraction)
+          $ contract edge
+    , morphism = sumModelAbstraction abstraction $ morphism edge
     }
 
+gotoSum ::
+  forall ap am bp bm.
+  Enumerable bp =>
+  HasParameterisedGenerator ap am =>
+  Abstraction ap am bp bm ->
+  Maybe (Morphism bp bm)
+gotoSum ProductAbstraction {} = Nothing
+gotoSum s@SumAbstraction {} =
+  Just $
+    Morphism
+      { name = "goto " <> abstractionName s
+      , match = Not $ Var $ propLabel s
+      , contract = clear >> add (propLabel s) >> addAll ((propertyAbstraction s #) <$> satisfiedBy)
+      , morphism =
+          const $
+            (sumModelAbstraction s #)
+              <$> genSatisfying @ap @am (All $ Var <$> satisfiedBy)
+      }
+
```

Comment:
Can you add a comment explaining what this does?

My understanding is it is allowing us to move between constructors. It's basically throwing everything away and generating from scratch in the target constructor.

An aside: We are doing this at quite a fine level of detail. It would be nice if we could have a way of generating ANY Left/Right then use the morphisms rather than adding all of these edges but that would change things a bit. I believe this came up in conversation before  - it might have been Maciej that mentioned it - the idea was to have the concept of a `Source` as well as the concept of a `Morphism`. If we had a smarter base generator that could pick from Left/Right sources before doing the graph traversal this might be more efficient. What do you think?

I suppose a `Source` would not have a contract - just a match clause - and it must generate something that matches. This would let us say `match = IsRight` for example. Then our base generator could be composed from a set of Sources that inject into each constructor of the Sum type.